domain/application/service/application.go

```diff
@@ -356,6 +370,197 @@ func (s *ApplicationService) validateStorageDirectives(ctx context.Context, mode
+}
+
+// Broker provides access to the k8s cluster to guery the scale
+// of a specified application.
+type Broker interface {
+	Application(string, caas.DeploymentType) caas.Application
+}
+
+// CAASUnitTerminating should be called by the CAASUnitTerminationWorker when
+// the agent receives a signal to exit. UnitTerminating will return how
+// the agent should shutdown.
+// We pass in a CAAS broker to get app details from the k8s cluster - we will probably
+// make it a service attribute once more use cases emerge.
+func (s *ApplicationService) CAASUnitTerminating(ctx context.Context, appName string, unitNum int, broker Broker) (bool, error) {
+	// TODO(sidecar): handle deployment other than statefulset
+	deploymentType := caas.DeploymentStateful
+	restart := true
+
+	switch deploymentType {
+	case caas.DeploymentStateful:
+		caasApp := broker.Application(appName, caas.DeploymentStateful)
+		appState, err := caasApp.State()
+		if err != nil {
+			return false, errors.Trace(err)
+		}
+		var scaleInfo application.ScaleState
+		err = s.st.RunAtomic(ctx, func(ctx domain.AtomicContext) error {
+			appID, err := s.st.GetApplicationID(ctx, appName)
+			if err != nil {
+				return errors.Trace(err)
+			}
+			scaleInfo, err = s.st.ApplicationScaleState(ctx, appID)
+			return errors.Trace(err)
+		})
+		if err != nil {
+			return false, errors.Trace(err)
+		}
+		if unitNum >= scaleInfo.Scale || unitNum >= appState.DesiredReplicas {
+			restart = false
+		}
+	case caas.DeploymentStateless, caas.DeploymentDaemon:
+		// Both handled the same way.
+		restart = true
+	default:
+		return false, errors.NotSupportedf("unknown deployment type")
+	}
+	return restart, nil
+}
+
+// SetApplicationScale sets the application's desired scale value, returning an error
+// satisfying [applicationerrors.ApplicationNotFound] if the application is not found.
+// This is used on CAAS models.
+func (s *ApplicationService) SetApplicationScale(ctx context.Context, appName string, scale int, force bool) error {
+	if scale < 0 {
+		return fmt.Errorf("application scale %d not valid%w", scale, errors.Hide(applicationerrors.ScaleChangeInvalid))
+	}
+	err := s.st.RunAtomic(ctx, func(ctx domain.AtomicContext) error {
+		appID, err := s.st.GetApplicationID(ctx, appName)
+		if err != nil {
+			return errors.Trace(err)
+		}
+		appScale, err := s.st.ApplicationScaleState(ctx, appID)
+		if err != nil {
+			return errors.Annotatef(err, "getting application scale state for app %q", appID)
+		}
+		s.logger.Tracef(
+			"SetScale DesiredScaleProtected %v, DesiredScale %v -> %v",
+			appScale.DesiredScaleProtected, appScale.Scale, scale,
+		)
+		if appScale.DesiredScaleProtected && !force && scale != appScale.Scale {
+			return fmt.Errorf(
+				"%w: SetScale(%d) without force while desired scale %d is not applied yet",
+				applicationerrors.ScaleChangeInvalid, scale, appScale.Scale)
+		}
+		return s.st.SetDesiredApplicationScale(ctx, appID, scale, force)
```

Comment:
Why is force passed to SetDesiredApplicationScale's `protected` argument?

Reply:
actually, after discussing with @wallyworld, DesiredScaleProtected is no longer needed.